### PR TITLE
Preventing prototype pollution vulnerabilities

### PR DESCRIPTION
**Pull request by AnastasiiaSvietlova**
## Goal

Preventing the `__proto__` , `constructor` or `prototype` property from being used as a section in `clear` and `add` methods of metadata-delegate.js

## Changeset

Added condition if `__proto__ || constructor || prototype` tries to pass as a section in the `clear` or `add` method, it will not be executed. 

## Testing

Created unit tests for checking 'constructor' and 'prototype'. 
it doesn't seem easy to check whether __proto__ keys can be overwritten

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,12 @@
 
 ### Changed
 
+- (metadata-delegate) Preventing prototype pollution vulnerabilities [#2115](https://github.com/bugsnag/bugsnag-js/pull/2115)
+
+## TBD
+
+### Changed
```
Comment (gingerbenw) [inline CHANGELOG.md:11]:
```suggestion
```

Duplicate header here

File: packages/core/test/metadata-delegate.test.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import { add, clear } from '../lib/metadata-delegate'
+
+// it doesn't seem easy or even impossible to check whether __proto__ keys can be overwritten
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/proto
+// so tests are only for prototype and constructor
+
+describe('metadata delegate', () => {
+  describe('clear', () => {
+    it('should not overwrite prototype keys', () => {
```
Comment (gingerbenw) [inline packages/core/test/metadata-delegate.test.js:9]: You could also use [.each](https://jestjs.io/docs/api#testeach) in this kind of circumstance, but just a suggestion!

**Comment by AnastasiiaSvietlova** [comment]
Yes, it was only for `clear` method. But I think it's a good point to do the same for `add`.

File: CHANGELOG.md [hunk 3]
```diff
@@ -4,6 +4,8 @@
 
 ### Changed
 
+- (metadata-delegate) Preventing prototype pollution vulnerabilities [#2115](https://github.com/bugsnag/bugsnag-js/pull/2115)
+
 - (plugin-interaction-breadcrumbs) Improved performance of click event breadcrumbs [#2094](https://github.com/bugsnag/bugsnag-js/pull/2094)
```
Comment (gingerbenw) [inline CHANGELOG.md:9]: Just a bit of formatting:

```suggestion
- (plugin-interaction-breadcrumbs) Improved performance of click event breadcrumbs [#2094](https://github.com/bugsnag/bugsnag-js/pull/2094)
- (metadata-delegate) Preventing prototype pollution vulnerabilities [#2115](https://github.com/bugsnag/bugsnag-js/pull/2115)
```

File: packages/core/test/metadata-delegate.test.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    it.each([
+      {
+        key: 'constructor',
+        expected: {}
+      },
+      {
+        key: 'prototype',
+        expected: {}
+      }
+    ])('should not add constructor or prototype keys', ({ key, expected }) => {
```
Comment (gingerbenw) [inline packages/core/test/metadata-delegate.test.js:18]:
```suggestion
    ])('should not add $key keys', ({ key, expected }) => {
```

File: packages/core/test/metadata-delegate.test.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            foo: 'bar'
+          }
+        },
+        expected: {
+          prototype: {
+            foo: 'bar'
+          }
+        }
+      }
+    ])('should not overwrite constructor or prototype keys', ({ key, state, expected }) => {
```
Comment (gingerbenw) [inline packages/core/test/metadata-delegate.test.js:53]:
```suggestion
    ])('should not overwrite $key keys', ({ key, state, expected }) => {
```

**Review by gingerbenw — APPROVED** [review]
Approved with a couple of minor suggestions 👍🏻

**Pull request merged**